File: src/main/java/app/application/utils/YoutubeIdExtractor.java
package app.application.utils;


import app.application.exception.InvalidPlaylistUrlException;
import app.application.exception.InvalidVideoUrlException;
import org.springframework.stereotype.Service;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.Map;

@Service
public class YoutubeIdExtractor {

    private final YoutubeUrlValidator youtubeUrlValidator;
    private final LinkParameterBuilder linkParameterBuilder;

    public YoutubeIdExtractor(YoutubeUrlValidator youtubeUrlValidator, LinkParameterBuilder linkParameterBuilder) {
        this.youtubeUrlValidator = youtubeUrlValidator;
        this.linkParameterBuilder = linkParameterBuilder;
    }

    public String getVideoIdFromLink(String videoLink) throws InvalidVideoUrlException {
        try {
            URL url = new URL(videoLink);
            Map<String, String> parameterMap = linkParameterBuilder.buildParameterMap(url.getQuery());
            return youtubeUrlValidator.isShortUrl(url.getHost()) ? url.getPath().substring(1) : parameterMap.get("v");
        } catch (MalformedURLException e) {
            throw new InvalidVideoUrlException(e, videoLink);
        }
    }

    public String getPlayListIdFromLink(String playlistLink) throws InvalidPlaylistUrlException {
        try {
            URL url = new URL(playlistLink);
            Map<String, String> parameterMap = linkParameterBuilder.buildParameterMap(url.getQuery());
            return parameterMap.getOrDefault("list", "");
        } catch (MalformedURLException e) {
            throw new InvalidPlaylistUrlException(e, playlistLink);
        }
    }

}

File: src/main/kotlin/app/application/utils/YoutubeIdExtractor.kt
package app.application.utils

import app.application.utils.YoutubeUrlValidator
import app.application.utils.LinkParameterBuilder
import kotlin.Throws
import app.application.exception.InvalidVideoUrlException
import java.net.MalformedURLException
import app.application.exception.InvalidPlaylistUrlException
import org.springframework.stereotype.Service
import java.net.URL

@Service
class YoutubeIdExtractor(private val youtubeUrlValidator: YoutubeUrlValidator, private val linkParameterBuilder: LinkParameterBuilder) {
    @Throws(InvalidVideoUrlException::class)
    fun getVideoIdFromLink(videoLink: String): String {
        return try {
            val url = URL(videoLink)
            val parameterMap = linkParameterBuilder.buildParameterMap(url.query)
            if (youtubeUrlValidator.isShortUrl(url.host)) url.path.substring(1) else parameterMap["v"]!!
        } catch (e: MalformedURLException) {
            throw InvalidVideoUrlException(e, videoLink)
        }
    }

    @Throws(InvalidPlaylistUrlException::class)
    fun getPlayListIdFromLink(playlistLink: String): String {
        return try {
            val url = URL(playlistLink)
            val parameterMap = linkParameterBuilder.buildParameterMap(url.query)
            parameterMap.getOrDefault("list", "")
        } catch (e: MalformedURLException) {
            throw InvalidPlaylistUrlException(e, playlistLink)
        }
    }
}